### PR TITLE
[rush] Overhaul of stream-collator pipeline

### LINKED ISSUE #2135: [heft] Rush stream-collator does not collate timing messages
## Please prefix the issue title with the project name i.e. [rush], [api-extractor] etc. ##

**Is this a feature or a bug?**

- [ ] Feature
- [x] Bug

**Please describe the actual behavior.**

In a recent CI build log for the Rush Stack repo, I noticed this output:

```
2020-08-21T22:43:03.3805464Z 44 of 89: [@microsoft/web-library-build] completed successfully in 11.58 seconds
2020-08-21T22:43:03.3806327Z [@rushstack/module-minifier-plugin] started
2020-08-21T22:43:03.3807673Z >>> @rushstack/module-minifier-plugin
2020-08-21T22:43:03.3808472Z heft test --clean --production
2020-08-21T22:43:03.4326428Z Using local Heft from /home/vsts/work/1/s/webpack/module-minifier-plugin/node_modules/@rushstack/heft
2020-08-21T22:43:03.4335074Z 
2020-08-21T22:43:04.3437086Z Project build folder is "/home/vsts/work/1/s/webpack/module-minifier-plugin"
2020-08-21T22:43:04.3463028Z Starting test
2020-08-21T22:43:04.3564802Z  ---- Clean started ---- 
2020-08-21T22:43:04.3582857Z Deleted 4 paths
2020-08-21T22:43:04.3586604Z  ---- Clean finished (2ms) ---- 
2020-08-21T22:43:04.3662648Z  ---- Compile started ---- 
2020-08-21T22:43:04.4026359Z [copy-static-assets] Copied 0 static assets in 0ms
2020-08-21T22:43:05.0702337Z [typescript] Using TypeScript version 3.5.3
2020-08-21T22:43:05.0703565Z [eslint] Using ESLint version 7.2.0
2020-08-21T22:43:13.5294701Z  ---- Compile finished (9162ms) ---- 
2020-08-21T22:43:13.5306272Z  ---- Bundle started ---- 
2020-08-21T22:43:14.1008070Z [api-extractor] Using API Extractor version 7.9.8
2020-08-21T22:43:15.4903071Z [api-extractor] Analysis will use the bundled TypeScript version 3.9.7
2020-08-21T22:43:17.8259727Z 45 of 89: [@rushstack/localization-plugin] completed successfully in 14.97 seconds
2020-08-21T22:43:17.8261394Z [api-extractor-lib2-test] started
2020-08-21T22:43:18.3640550Z  ---- Bundle finished (4831ms) ---- 
2020-08-21T22:43:18.3642243Z  ---- Test started ---- 
2020-08-21T22:43:18.7891086Z �[0m
2020-08-21T22:43:18.7895344Z Run start. 2 test suites
2020-08-21T22:43:18.7914168Z START src/test/RehydrateAsset.test.ts
2020-08-21T22:43:20.4355031Z PASS src/test/RehydrateAsset.test.ts (duration: ?, 8 passed, 0 failed)
2020-08-21T22:43:20.4362599Z START src/test/MinifiedIdentifier.test.ts
2020-08-21T22:43:20.6307696Z PASS src/test/MinifiedIdentifier.test.ts (duration: ?, 2 passed, 0 failed)
```

The "Bundle started" and "Bundle finished" lines are clearly from the same `module-minifier-plugin` log. But there's Rush output is mixed in the middle of those lines.

**What is the expected behavior?**

The [stream-collator](https://www.npmjs.com/package/@rushstack/stream-collator) is a showcase feature of Rush. It collates log output, enabling developers to view realtime output from parallel builds, without having to awkwardly prefix each log line with an associated project name.

Whereas when the output is interleaved as above, it is difficult to interpret.  Suppose a unit test failed:

```
2020-08-21T22:43:17.8261394Z [api-extractor-lib2-test] started
2020-08-21T22:43:18.3640550Z  ---- Bundle finished (4831ms) ---- 
2020-08-21T22:43:18.3642243Z  ---- Test started ---- 
2020-08-21T22:43:18.7891086Z �[0m
2020-08-21T22:43:18.7895344Z Run start. 2 test suites
2020-08-21T22:43:18.7914168Z START src/test/RehydrateAsset.test.ts
2020-08-21T22:43:18.7914168Z FAIL: Some problem with RehydrateAsset
```

Which project is this referring to?  Because Rush is not collating properly, the above error would seem to apply to `api-extractor-lib2-test`, but actually it is referring to `module-minifier-plugin`.

**Remarks**

We recently switched from `gulp-core-build` to Heft --  maybe Heft is somehow directly writing to the terminal's STDOUT, preventing Rush from capturing and collating the stream?

### PR DESCRIPTION
**Pull request by octogonz**
Fixes #2135

Along the way I also redesigned the `rush build` output to be more readable.

We also introduced a `@rushstack/terminal` package that implements various transforms for manipulating stderr/stdout as streams. This solves a scalability issue where `rush build` was storing the entire build output in a huge text string in RAM.

### CODE REVIEW

File: apps/rush-lib/src/logic/taskRunner/TaskRunner.ts [hunk 1]
```diff
@@ -2,26 +2,28 @@
...
-import { TaskError } from './TaskError';
+import { IBuilderContext } from './BaseBuilder';
 
 export interface ITaskRunnerOptions {
   quietMode: boolean;
   parallelism: string | undefined;
   changedProjectsOnly: boolean;
   allowWarningsInSuccessfulBuild: boolean;
-  terminal?: Terminal;
+  destination?: TerminalWritable;
```
Comment (iclanton) [inline apps/rush-lib/src/logic/taskRunner/TaskRunner.ts:26]: "destination" is confusing in this context. It kinda sounds like this is where build output should go.

**Review by iclanton — APPROVED** [review]
This looks good overall. One thing the new output is missing is a way to see which projects are currently building, as you now just find out when they finish. If you have a few long-running processes, it can feel like the log has stalled. A view that prints the names of the projects as they start and then updates those lines as they finish would be the most useful view IMO.

Something like:

```
==[ @microsoft/rushell ]============================{ building }=[ 81 of 95 ]== (blue)
==[ heft-oldest-compiler-test ]====================={ building }=[ 82 of 95 ]== (blue)
==[ heft-node-everything-test ]====================={ building }=[ 83 of 95 ]== (blue)
```

(and then after `@microsoft/rushell` finishes)

```
==[ @microsoft/rushell ]================================{ done }=[ 81 of 95 ]== (green)
==[ heft-oldest-compiler-test ]====================={ building }=[ 82 of 95 ]== (blue)
==[ heft-node-everything-test ]====================={ building }=[ 83 of 95 ]== (blue)
```

(and then if `heft-oldest-compiler-test` failed and some other stuff happened)

```
==[ @microsoft/rushell ]================================{ done }=[ 81 of 95 ]== (green)
==[ heft-oldest-compiler-test ]======================={ failed }=[ 82 of 95 ]== (red)
==[ heft-node-everything-test ]====================={ building }=[ 83 of 95 ]== (blue)
==[ heft-node-jest-test ]============================{ skipped }=[ 84 of 95 ]== (gray)
==[ heft-rsc-test ]================================={ warnings }=[ 85 of 95 ]== (yellow)
```

There is also a LOT of whitespace in the log. Can we get rid of the newline after each heading?

**Comment by octogonz** [comment]
> A view that prints the names of the projects as they start and then updates those lines as they finish would be the most useful view IMO.
> 
> Something like:
> 
> ```
> ==[ @microsoft/rushell ]============================{ building }=[ 81 of 95 ]== (blue)
> ==[ heft-oldest-compiler-test ]====================={ building }=[ 82 of 95 ]== (blue)
> ==[ heft-node-everything-test ]====================={ building }=[ 83 of 95 ]== (blue)
> ```
> 
> (and then after `@microsoft/rushell` finishes)
> 
> ```
> ==[ @microsoft/rushell ]================================{ done }=[ 81 of 95 ]== (green)
> ==[ heft-oldest-compiler-test ]====================={ building }=[ 82 of 95 ]== (blue)
> ==[ heft-node-everything-test ]====================={ building }=[ 83 of 95 ]== (blue)
> ```

Sounds great, but unfortunately it's out of scope for this PR to introduce ANSI codes that move the cursor around and overwrite previously written text. Also, this sort of painting will not be usable in a CI log, so even if we do implement it (which definitely does fit the mission of `@rushstack/terminal`), we would still need a design for contexts that don't support cursor moving.

File: apps/rush-lib/src/logic/taskRunner/TaskRunner.ts (hunk 1 above)
Comment (octogonz) [reply apps/rush-lib/src/logic/taskRunner/TaskRunner.ts:26]: agreed

**Review by KevinTCoughlin — APPROVED** [review]

**Comment by octogonz** [comment]
> There is also a LOT of whitespace in the log. Can we get rid of the newline after each heading?

Thinking about this, the newline really is needed when verbose stderr/stdout is being written in each section.  

But if you do `rush build` WITHOUT `--verbose`, then we can remove one of the newlines and it still looks okay,  even if there is stderr output.  I pushed a commit with this change.

**Comment by octogonz** [comment]
This was released with Rush 5.34.1.

**Pull request merged**